File: leetcode/1325-delete-leaves-with-a-given-value/Solution.java
import java.util.*;

/**
 * Definition for a binary tree node.
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    /**
     * 删除二叉树中所有值为 target 的叶子节点
     * 使用后序遍历，先处理子节点再处理当前节点
     *
     * 时间复杂度：O(n)，每个节点访问一次
     * 空间复杂度：O(h)，递归栈深度为树的高度
     */
    public TreeNode removeLeafNodes(TreeNode root, int target) {
        if (root == null) {
            return null;
        }

        // 后序遍历：先递归处理左右子树
        root.left = removeLeafNodes(root.left, target);
        root.right = removeLeafNodes(root.right, target);

        // 如果当前节点变成了叶子节点且值等于 target，删除它
        if (root.left == null && root.right == null && root.val == target) {
            return null;
        }

        return root;
    }

    /**
     * 从层序列表构建二叉树
     */
    public static TreeNode buildTree(Integer[] values) {
        if (values == null || values.length == 0 || values[0] == null) {
            return null;
        }
        TreeNode root = new TreeNode(values[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;
        while (!queue.isEmpty() && i < values.length) {
            TreeNode node = queue.poll();
            if (i < values.length && values[i] != null) {
                node.left = new TreeNode(values[i]);
                queue.offer(node.left);
            }
            i++;
            if (i < values.length && values[i] != null) {
                node.right = new TreeNode(values[i]);
                queue.offer(node.right);
            }
            i++;
        }
        return root;
    }

    /**
     * 将二叉树转为层序列表
     */
    public static List<Integer> treeToList(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            if (node != null) {
                result.add(node.val);
                queue.offer(node.left);
                queue.offer(node.right);
            } else {
                result.add(null);
            }
        }
        // 去除末尾的 null
        while (!result.isEmpty() && result.get(result.size() - 1) == null) {
            result.remove(result.size() - 1);
        }
        return result;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // 测试用例 1: [1,2,3,2,null,2,4], target=2 -> [1,null,3,null,4]
        TreeNode root1 = buildTree(new Integer[]{1, 2, 3, 2, null, 2, 4});
        TreeNode result1 = sol.removeLeafNodes(root1, 2);
        System.out.println("Test 1: " + treeToList(result1));

        // 测试用例 2: [1,3,3,3,2], target=3 -> [1,3,null,null,2]
        TreeNode root2 = buildTree(new Integer[]{1, 3, 3, 3, 2});
        TreeNode result2 = sol.removeLeafNodes(root2, 3);
        System.out.println("Test 2: " + treeToList(result2));

        // 测试用例 3: [1,2,null,2,null,2], target=2 -> [1]
        TreeNode root3 = buildTree(new Integer[]{1, 2, null, 2, null, 2});
        TreeNode result3 = sol.removeLeafNodes(root3, 2);
        System.out.println("Test 3: " + treeToList(result3));
    }
}

File: leetcode/1325-delete-leaves-with-a-given-value/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def removeLeafNodes(self, root: Optional[TreeNode], target: int) -> Optional[TreeNode]:
        # 时间复杂度：O(n)，每个节点访问一次
        # 空间复杂度：O(h)，递归栈深度为树的高度
        if not root:
            return None

        # 后序遍历：先处理左右子树
        root.left = self.removeLeafNodes(root.left, target)
        root.right = self.removeLeafNodes(root.right, target)

        # 如果当前节点变成了叶子且值等于 target，删除它
        if not root.left and not root.right and root.val == target:
            return None

        return root


def build_tree_from_list(values, index=0):
    """从层序列表构建二叉树"""
    if index >= len(values) or values[index] is None:
        return None
    node = TreeNode(values[index])
    node.left = build_tree_from_list(values, 2 * index + 1)
    node.right = build_tree_from_list(values, 2 * index + 2)
    return node


def tree_to_list(root):
    """将二叉树转为层序列表"""
    if not root:
        return []
    result = []
    queue = [root]
    while queue:
        node = queue.pop(0)
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    # 去除末尾的 None
    while result and result[-1] is None:
        result.pop()
    return result


if __name__ == "__main__":
    sol = Solution()

    # 测试用例 1: [1,2,3,2,null,2,4], target=2 -> [1,null,3,null,4]
    root = build_tree_from_list([1, 2, 3, 2, None, 2, 4])
    result = sol.removeLeafNodes(root, 2)
    print("Test 1:", tree_to_list(result))  # [1, None, 3, None, 4]

    # 测试用例 2: [1,3,3,3,2], target=3 -> [1,3,null,null,2]
    root = build_tree_from_list([1, 3, 3, 3, 2])
    result = sol.removeLeafNodes(root, 3)
    print("Test 2:", tree_to_list(result))  # [1, 3, None, None, 2]

    # 测试用例 3: [1,2,null,2,null,2], target=2 -> [1]
    root = build_tree_from_list([1, 2, None, 2, None, 2])
    result = sol.removeLeafNodes(root, 2)
    print("Test 3:", tree_to_list(result))  # [1]
